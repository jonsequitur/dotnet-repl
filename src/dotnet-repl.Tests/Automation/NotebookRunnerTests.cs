using System;
using System.CommandLine.IO;
using System.CommandLine.Parsing;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using Assent;
using Automation;
using dotnet_repl.Tests.Utility;
using FluentAssertions;
using Microsoft.DotNet.Interactive.CSharp;
using Microsoft.DotNet.Interactive.Documents;
using Microsoft.DotNet.Interactive.Documents.Jupyter;
using Microsoft.DotNet.Interactive.Formatting;
using Pocket;
using Spectre.Console;
using TRexLib;
using Xunit;

namespace dotnet_repl.Tests.Automation;

public class NotebookRunnerTests : IDisposable
{
    private readonly CompositeDisposable _disposables = new();
    private readonly StringWriter _writer;
    private readonly string _directory = Path.GetDirectoryName(PathUtility.PathToCurrentSourceFile());
    private readonly Parser _parser;
    private readonly TestConsole console = new();

    private readonly Configuration _assentConfiguration =
        new Configuration()
            .UsingExtension("json")
            .SetInteractive(Debugger.IsAttached);

    public NotebookRunnerTests()
    {
        var ansiConsole = AnsiConsole.Create(new AnsiConsoleSettings
        {
            Ansi = AnsiSupport.Yes,
            Interactive = InteractionSupport.Yes,
            Out = new AnsiConsoleOutput(_writer = new StringWriter())
        });

        _parser = CommandLineParser.Create(ansiConsole, registerForDisposal: d => _disposables.Add(d));

        _disposables.Add(_writer);
    }

    public void Dispose() => _disposables.Dispose();

    [Fact]
    public async Task When_an_ipynb_is_run_and_no_error_is_produced_then_the_exit_code_is_0()
    {
        var result = await _parser.InvokeAsync($"--notebook \"{_directory}/succeed.ipynb\" --exit-after-run", console);

        var output = _writer.ToString();
        console.Error.ToString().Should().BeEmpty();
        result.Should().Be(0);
        output.Should().Contain("Success!");
    }

    [Fact]
    public async Task When_an_ipynb_is_run_and_an_error_is_produced_from_a_cell_then_the_exit_code_is_2()
    {
        var result = await _parser.InvokeAsync($"--notebook \"{_directory}/fail.ipynb\" --exit-after-run", console);

        var output = _writer.ToString();
        console.Error.ToString().Should().BeEmpty();
        result.Should().Be(2);
        output.Should().Contain("Oops!");
    }

    [Fact]
    public async Task Notebook_runner_produces_expected_output()
    {
        using var kernel = KernelBuilder.CreateKernel();

        var runner = new NotebookRunner(kernel);

        var notebookFile = Path.Combine(_directory, "VS Code outputs.ipynb");

        var expectedContent = await File.ReadAllTextAsync(notebookFile);

        var inputDoc = Notebook.Parse(expectedContent, new(kernel.ChildKernels.Select(k => new KernelName(k.Name)).ToArray()));

        var resultDoc = await runner.RunNotebookAsync(inputDoc);

        NormalizeMetadata(resultDoc);

        var resultContent = resultDoc.Serialize();

        this.Assent(resultContent, _assentConfiguration);
    }

    [Fact]
    public async Task ASCII_escape_sequences_do_not_cause_XML_parse_problems()
    {
        using var kernel = new CSharpKernel();
        
        var runner = new NotebookRunner(kernel);

        var inputDoc = new InteractiveDocument
        {
            new InteractiveDocumentElement("123.Display(\"unknown/MIMEtype\");", "csharp")
        };
        var outputDoc = await runner.RunNotebookAsync(inputDoc);

        var xml = outputDoc.ToTestOutputDocumentXml();

        TestOutputDocumentParser.Parse(xml);



        // TODO (ASCII_escape_sequences_do_not_cause_XML_parse_problems) write test
        throw new NotImplementedException();
    }

    private void NormalizeMetadata(InteractiveDocument document)
    {
        foreach (var element in document.Elements)
        {
            if (element.Metadata is { })
            {
                if (element.Metadata.ContainsKey("dotnet_repl_cellExecutionStartTime"))
                {
                    element.Metadata["dotnet_repl_cellExecutionStartTime"] = DateTimeOffset.MinValue;
                }

                if (element.Metadata.ContainsKey("dotnet_repl_cellExecutionEndTime"))
                {
                    element.Metadata["dotnet_repl_cellExecutionEndTime"] = DateTimeOffset.MinValue;
                }
            }
        }
    }
}