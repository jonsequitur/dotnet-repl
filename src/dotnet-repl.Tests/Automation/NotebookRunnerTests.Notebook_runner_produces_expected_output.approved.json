{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "dotnet_repl_cellExecutionStartTime": "0001-01-01T00:00:00+00:00",
    "dotnet_repl_cellExecutionEndTime": "0001-01-01T00:00:00+00:00",
    "dotnet_interactive": {
     "language": "csharp"
    },
    "polyglot_notebook": {
     "kernelName": "csharp"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "output\r\n"
     ]
    }
   ],
   "source": [
    "Console.Out.WriteLine(\"output\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "dotnet_repl_cellExecutionStartTime": "0001-01-01T00:00:00+00:00",
    "dotnet_repl_cellExecutionEndTime": "0001-01-01T00:00:00+00:00",
    "dotnet_interactive": {
     "language": "csharp"
    },
    "polyglot_notebook": {
     "kernelName": "csharp"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "error\r\n"
     ]
    }
   ],
   "source": [
    "Console.Error.WriteLine(\"error\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "dotnet_repl_cellExecutionStartTime": "0001-01-01T00:00:00+00:00",
    "dotnet_repl_cellExecutionEndTime": "0001-01-01T00:00:00+00:00",
    "dotnet_interactive": {
     "language": "csharp"
    },
    "polyglot_notebook": {
     "kernelName": "csharp"
    }
   },
   "outputs": [
    {
     "ename": "Error",
     "evalue": "System.Exception: oops!\r\n   at Submission#3.<<Initialize>>d__0.MoveNext()\r\n--- End of stack trace from previous location ---\r\n   at Microsoft.CodeAnalysis.Scripting.ScriptExecutionState.RunSubmissionsAsync[TResult](ImmutableArray`1 precedingExecutors, Func`2 currentExecutor, StrongBox`1 exceptionHolderOpt, Func`2 catchExceptionOpt, CancellationToken cancellationToken)",
     "output_type": "error",
     "traceback": [
      "   at Submission#3.<<Initialize>>d__0.MoveNext()",
      "--- End of stack trace from previous location ---",
      "   at Microsoft.CodeAnalysis.Scripting.ScriptExecutionState.RunSubmissionsAsync[TResult](ImmutableArray`1 precedingExecutors, Func`2 currentExecutor, StrongBox`1 exceptionHolderOpt, Func`2 catchExceptionOpt, CancellationToken cancellationToken)"
     ]
    }
   ],
   "source": [
    "throw new Exception(\"oops!\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "dotnet_repl_cellExecutionStartTime": "0001-01-01T00:00:00+00:00",
    "dotnet_repl_cellExecutionEndTime": "0001-01-01T00:00:00+00:00",
    "dotnet_interactive": {
     "language": "csharp"
    },
    "polyglot_notebook": {
     "kernelName": "csharp"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "A return value"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\"A return value\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "dotnet_repl_cellExecutionStartTime": "0001-01-01T00:00:00+00:00",
    "dotnet_repl_cellExecutionEndTime": "0001-01-01T00:00:00+00:00",
    "dotnet_interactive": {
     "language": "csharp"
    },
    "polyglot_notebook": {
     "kernelName": "csharp"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"dni-plaintext\"><pre>An HTML display value</pre></div><style>\r\n",
       ".dni-code-hint {\r\n",
       "    font-style: italic;\r\n",
       "    overflow: hidden;\r\n",
       "    white-space: nowrap;\r\n",
       "}\r\n",
       ".dni-treeview {\r\n",
       "    white-space: nowrap;\r\n",
       "}\r\n",
       ".dni-treeview td {\r\n",
       "    vertical-align: top;\r\n",
       "    text-align: start;\r\n",
       "}\r\n",
       "details.dni-treeview {\r\n",
       "    padding-left: 1em;\r\n",
       "}\r\n",
       "table td {\r\n",
       "    text-align: start;\r\n",
       "}\r\n",
       "table tr { \r\n",
       "    vertical-align: top; \r\n",
       "    margin: 0em 0px;\r\n",
       "}\r\n",
       "table tr td pre \r\n",
       "{ \r\n",
       "    vertical-align: top !important; \r\n",
       "    margin: 0em 0px !important;\r\n",
       "} \r\n",
       "table th {\r\n",
       "    text-align: start;\r\n",
       "}\r\n",
       "</style>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\"An HTML display value\".Display(\"text/html\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "dotnet_repl_cellExecutionStartTime": "0001-01-01T00:00:00+00:00",
    "dotnet_repl_cellExecutionEndTime": "0001-01-01T00:00:00+00:00",
    "dotnet_interactive": {
     "language": "csharp"
    },
    "polyglot_notebook": {
     "kernelName": "csharp"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "A plaintext display value"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\"A plaintext display value\".Display(\"text/plain\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "dotnet_repl_cellExecutionStartTime": "0001-01-01T00:00:00+00:00",
    "dotnet_repl_cellExecutionEndTime": "0001-01-01T00:00:00+00:00",
    "dotnet_interactive": {
     "language": "csharp"
    },
    "polyglot_notebook": {
     "kernelName": "csharp"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "A plaintext display value (in a multi-output cell)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div class=\"dni-plaintext\"><pre>An HTML display value (in a multi-output cell)</pre></div><style>\r\n",
       ".dni-code-hint {\r\n",
       "    font-style: italic;\r\n",
       "    overflow: hidden;\r\n",
       "    white-space: nowrap;\r\n",
       "}\r\n",
       ".dni-treeview {\r\n",
       "    white-space: nowrap;\r\n",
       "}\r\n",
       ".dni-treeview td {\r\n",
       "    vertical-align: top;\r\n",
       "    text-align: start;\r\n",
       "}\r\n",
       "details.dni-treeview {\r\n",
       "    padding-left: 1em;\r\n",
       "}\r\n",
       "table td {\r\n",
       "    text-align: start;\r\n",
       "}\r\n",
       "table tr { \r\n",
       "    vertical-align: top; \r\n",
       "    margin: 0em 0px;\r\n",
       "}\r\n",
       "table tr td pre \r\n",
       "{ \r\n",
       "    vertical-align: top !important; \r\n",
       "    margin: 0em 0px !important;\r\n",
       "} \r\n",
       "table th {\r\n",
       "    text-align: start;\r\n",
       "}\r\n",
       "</style>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\"A plaintext display value (in a multi-output cell)\".Display(\"text/plain\");\n",
    "\"An HTML display value (in a multi-output cell)\".Display(\"text/html\");"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".NET (C#)",
   "language": "C#",
   "name": ".net-csharp"
  },
  "language_info": {
   "file_extension": ".cs",
   "mimetype": "text/x-csharp",
   "name": "C#",
   "pygments_lexer": "csharp",
   "version": "10.0"
  },
  "polyglot_notebook": {
   "defaultKernelName": "csharp",
   "items": [
    {
     "name": "csharp"
    }
   ]
  },
  "dotnet_interactive": {
   "defaultKernelName": "csharp",
   "items": [
    {
     "name": "csharp"
    }
   ]
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}